Fix SELECT * REPLACE behavior in WHERE clause with new analyzer

xiaohuanlin opened the pull request:
Fix SELECT * REPLACE behavior in WHERE clause with new analyzer

This commit fixes issue #85313 where SELECT * REPLACE with WHERE clause on the same column returned incorrect results with the new analyzer. For example:

SELECT * REPLACE(a + 10 AS b) FROM test WHERE b = 11 was returning empty results instead of matching rows where the replaced expression equals 11.

The root cause was that the new analyzer processes SELECT * REPLACE and WHERE clause sequentially, causing WHERE conditions to see original column values instead of replaced values. The old analyzer processes both together correctly.

The fix captures REPLACE transformer mappings during resolveMatcher processing and immediately applies inline identifier replacement to WHERE and HAVING clauses using lambda functions.

 ### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes into CHANGELOG.md):
- Fixed incorrect SELECT * REPLACE behavior in WHERE clause with new analyzer when filtering on replaced columns.

xiaohuanlin committed:
Fix SELECT * REPLACE behavior in WHERE clause with new analyzer

xiaohuanlin commented on src/Analyzer/Resolve/QueryAnalyzer.cpp:
```diff
@@ -2200,6 +2206,62 @@ ProjectionNames QueryAnalyzer::resolveMatcher(QueryTreeNodePtr & matcher_node, I
+                    query_node->getWhere() = where_node;
+                }
+
+                if (query_node->hasHaving())
```
Sure. I am checking

xiaohuanlin committed:
Extend SELECT * REPLACE fix to all SQL clauses with comprehensive tests

- Fix column reference issues in WHERE, PREWHERE, ORDER BY, GROUP BY, HAVING, LIMIT BY, WINDOW clauses
- Replace original test with comprehensive 17-scenario test suite
- Ensure consistent behavior between new and old analyzer

xiaohuanlin committed:
Fix WINDOW clause projection and add comprehensive test coverage

antaljanosbenjamin commented on tests/queries/0_stateless/03631_select_replace_comprehensive.sql:
```diff
@@ -0,0 +1,137 @@
+
+-- Test 4: LIMIT BY clause - should group by replaced values
+SELECT '=== LIMIT BY clause test ===';
+SELECT * REPLACE(a AS b) FROM test_replace_main ORDER BY id LIMIT 1 BY b;
```
I think there is no repeated values, so `LIMIT 1 BY b` is not tested. Maybe `LIMIT 1 BY b < 200`?

xiaohuanlin replied:
Fixed

antaljanosbenjamin commented on tests/queries/0_stateless/03631_select_replace_comprehensive.sql:
```diff
@@ -0,0 +1,137 @@
+SELECT category, sum(value) as total
+FROM (SELECT * REPLACE(value * 10 AS value) FROM test_group_by)
+GROUP BY category
+ORDER BY category;
```
Didn't want to use a replaced expression in `GROUP BY`?

xiaohuanlin committed:
Improve SELECT * REPLACE test coverage per reviewer feedback

Address review comments from @antaljanosbenjamin:

1. LIMIT BY test enhancement:
   - Add test data with duplicate values to properly test LIMIT BY grouping
   - Use expression 'b < 200' to test replaced column in LIMIT BY condition

2. GROUP BY test expansion:
   - Preserve subquery test (Test 16) - validates replaced values in subquery work with outer GROUP BY
   - Add direct usage test (Test 17) - validates replaced column can be used directly in GROUP BY clause

3. HAVING test expansion:
   - Preserve subquery test (Test 18) - validates replaced values in subquery work with outer HAVING
   - Add direct usage test (Test 19) - validates replaced column can be used directly in HAVING clause

Test coverage now includes both subquery patterns (important for nested query validation)
and direct usage patterns (requested by reviewer) for GROUP BY and HAVING clauses.

xiaohuanlin committed:
Clean up code comments and remove QUALIFY test

xiaohuanlin commented on tests/queries/0_stateless/03631_select_replace_comprehensive.sql:
```diff
@@ -0,0 +1,137 @@
+SELECT category, sum(amount) as total
+FROM (SELECT * REPLACE(amount + 100 AS amount) FROM test_having)
+GROUP BY category
+HAVING total > 200
```
Same as above

antaljanosbenjamin commented on tests/queries/0_stateless/03631_select_replace_comprehensive.reference:
```diff
@@ -14,9 +14,8 @@
-2	300	300	beta
-3	200	200	gamma
+1	100	100
+2	100	300
```
I think I don't understand this output. We have this table:
```
   +----+-----+-----+
   | id |   a |   b |
   +----+-----+-----+
1. |  1 | 100 | 200 |
   +----+-----+-----+
2. |  2 | 100 | 300 |
   +----+-----+-----+
3. |  3 | 200 | 400 |
   +----+-----+-----+
4. |  4 | 200 | 500 |
   +----+-----+-----+
```

And this query:
```
SELECT * REPLACE(a AS b) FROM test_limit_by ORDER BY id LIMIT 1 BY b < 200;
```

If I understood correctly, then `REPLACE(a AS b)` will replace all values of `b` with the values of `a`, so basically we would see this table:
```
   +----+-----+-----+
   | id |   a |   b |
   +----+-----+-----+
1. |  1 | 100 | 100 |
   +----+-----+-----+
2. |  2 | 100 | 100 |
   +----+-----+-----+
3. |  3 | 200 | 200 |
   +----+-----+-----+
4. |  4 | 200 | 200 |
   +----+-----+-----+
```
And because of `ORDER BY id LIMIT BY b < 200` we should see one row where `b` is smaller than `200` and one row where `b` isn't. It returns the first two rows (based on the `id`). I am also confused how the returned values (`100` vs `300`) can differ for the second row when `b` was replaced by `a`.

Could you please help me to understand what am I missing?

antaljanosbenjamin replied:
Ahh, sorry, I don't know why I commented on the old version.

antaljanosbenjamin reviewed (commented):
One question only to be sure we are not making a mistake here. Apart from that, it looks good.

antaljanosbenjamin commented:
Failed test is caused by https://github.com/ClickHouse/ClickHouse/issues/87655. Merging PR.
